### PR TITLE
Add custom color support to project folders

### PR DESCRIPTION
**Pull request by snibk**
This PR addresses https://github.com/godotengine/godot-proposals/issues/768 and adds the ability to give folders in a project a custom color when viewed in the FileSystem dock. Child files & folders will inherit their parent folder's color if they don't have one set.

As of b98981e:

image

image

`project.godot`:

image

### CODE REVIEW

**Comment by DennisGHUA** [comment]
I think the introduction of background colors has notably improved the efficiency of locating assets and folders. This enhancement is exemplified by the darker background color now visible under the "another_child" folder when opened. Prior to this update, Godot lacked a distinct differentiation between opened and closed folders. It's worth noting that the absence of open and closed folder icons in Godot remains a separate topic for consideration. Nevertheless, it might be beneficial to fine-tune the brightness of the background colors slightly to achieve an even more polished visual effect.

File: editor/filesystem_dock.cpp [hunk 1]
```diff
@@ -182,16 +182,38 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
 	// Create a tree item for the subdirectory.
 	TreeItem *subdirectory_item = tree->create_item(p_parent);
 	String dname = p_dir->get_name();
+	String lpath = p_dir->get_path();
+
 	if (dname.is_empty()) {
 		dname = "res://";
 	}
 
+	// Set custom folder color (if applicable)
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:191]:
```suggestion
	// Set custom folder color (if applicable).
```
[Comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide) nitpick

File: editor/filesystem_dock.cpp [hunk 2]
```diff
@@ -2711,6 +2742,54 @@ void FileSystemDock::_get_drag_target_folder(String &target, bool &target_favori
 	}
 }
 
+void FileSystemDock::_folder_color_index_pressed(int p_index) {
+	Dictionary settings = ProjectSettings::get_singleton()->get_setting("file_customization/folder_colors");
+	Variant chosen_color_name = folder_colors_menu->get_item_metadata(p_index);
+	Vector<String> selected;
+
+	// Get all selected folders based on whether the files panel or tree panel is currently focused, update folder color
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:2750]:
```suggestion
	// Get all selected folders based on whether the files panel or tree panel is currently focused, update folder color.
```

File: editor/filesystem_dock.cpp [hunk 3]
```diff
@@ -2711,6 +2742,54 @@ void FileSystemDock::_get_drag_target_folder(String &target, bool &target_favori
...
+	} else {
+		TreeItem *tree_selected = tree->get_root();
+		tree_selected = tree->get_next_selected(tree_selected);
+		while (tree_selected) {
+			selected.push_back(tree_selected->get_metadata(0));
+			tree_selected = tree->get_next_selected(tree_selected);
+		}
+	}
+
+	// Update dictionary with new folder colors
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:2765]:
```suggestion
	// Update dictionary with new folder colors.
```

File: editor/filesystem_dock.cpp [hunk 4]
```diff
@@ -2711,6 +2742,54 @@ void FileSystemDock::_get_drag_target_folder(String &target, bool &target_favori
...
+		String fpath = selected[i];
+
+		if (chosen_color_name) {
+			settings[fpath] = chosen_color_name;
+		} else {
+			settings.erase(fpath);
+		}
+	}
+
+	// Update file list parent dir manually
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:2776]:
```suggestion
	// Update file list parent dir manually.
```

File: editor/filesystem_dock.cpp [hunk 5]
```diff
@@ -182,16 +182,38 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
...
 	subdirectory_item->set_icon(0, get_theme_icon(SNAME("Folder"), SNAME("EditorIcons")));
-	subdirectory_item->set_icon_modulate(0, get_theme_color(SNAME("folder_icon_color"), SNAME("FileDialog")));
+	if (has_custom_color) {
+		subdirectory_item->set_icon_modulate(0, custom_color);
+		subdirectory_item->set_custom_bg_color(0, custom_color.darkened(0.6));
+	} else {
+		TreeItem *parent = subdirectory_item->get_parent();
+		if (parent) {
+			Color parent_bg_color = parent->get_custom_bg_color(0);
+			if (parent_bg_color != custom_color) { // a.k.a parent_bg_color != Color()
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:206]:
```suggestion
			if (parent_bg_color != Color()) {
```
I'd simplify this condition, as it's always equivalent, as per your comment

**Comment by the-sink** [comment]
@Calinou Yep, agreed - Reduced opacity of the background color instead of darkening it looks better, with the added benefit of fixing the highlight issue:
image

---

I also added a few additional color options and changed them all to more saturated versions that exist in the `editor_themes.cpp` color conversion list.
image


Many areas of the editor (i.e. icons, accent lines, etc) use lighter, more pastel-like colors so I initially tried to keep them closer to that appearance, but personally I like the more saturated colors in this context. Interested in what others think though.

File: editor/filesystem_dock.cpp (hunk 1 above)
Comment (the-sink) [reply editor/filesystem_dock.cpp]: Fixed in 24487ac1eb9c77a4edbd976bbcdd31d9d082b1c3. Thanks!

**Comment by DennisGHUA** [comment]
@the-sink Currently, I believe that opting for lighter, more pastel-like colors for the background would create a slightly less distracting environment.

**Comment by Mickeon** [comment]
I like the way more saturated colors for the folders. If they are coloured, it means the user finds them noteworthy, somewhat. 

Still not convinced on the background color. It's quite distracting, I feel. Although, it does grant the benefit of being able to discern under what folder everything is much more easily.

I do have a suggestion: what if only the nested elements inside the folder had the background color (still toned down a notch)?

A mock-up:
image

**Comment by Mickeon** [comment]
Editor settings should always be a last resort, I feel. It's worth choosing the most favourable default before considering options.

**Comment by the-sink** [comment]
This would require a separate PR, but another idea; what if the Tree relationship lines matched folder colors instead? An option could be added to the Tree object for path lines to inherit the icon modulate color of their parent items.

image


This still makes the folder hierarchy more visually clear to users, I think.

We'll also need to think about colorblindness accessibility & potentially having different patterns or shapes for these color options, however I'm not sure if that should be a part of this PR or a future one.

**Comment by Norrox** [comment]
Great work on this!
The more saturated folders look fine for me.
I actually prefer a more saturated background color for the primary folder. The stronger background helps in quickly identifying and differentiating the folders. While I recognize the concern about potential distractions, I feel that a bolder color, especially for the main folder, adds a layer of emphasis that can be beneficial. However, I’m on board with the idea of giving nested elements a more subtle shade. This can provide the right contrast and hierarchy between primary and nested folders. And for those who might find colors distracting, there’s always the option not to use them at all.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 4714e95896c8db02616ea6ec7f9aff92dec1cae4), it works as expected.

**Testing project:** [test_pr_80440.zip](https://github.com/godotengine/godot/files/12327104/test_pr_80440.zip)

The highlight colors look good to me now:

Default | Light | Black (OLED)
-|-|-
Screenshot_20230812_120437 | Screenshot_20230812_120448 | Screenshot_20230812_120427

We could consider adjusting the highlight colors later, but I think this is better left for a future PR after people start using this feature in their projects.

File: editor/filesystem_dock.cpp [hunk 6]
```diff
@@ -3319,6 +3424,17 @@ FileSystemDock::FileSystemDock() {
...
+	folder_colors = HashMap<String, Color>();
+	folder_colors["red"] = Color("#ff4545");
+	folder_colors["orange"] = Color("#ff8f45");
+	folder_colors["yellow"] = Color("#ffe345");
+	folder_colors["green"] = Color("#80ff45");
+	folder_colors["teal"] = Color("#45ffa2");
+	folder_colors["blue"] = Color("#45d7ff");
+	folder_colors["purple"] = Color("#8045ff");
+	folder_colors["pink"] = Color("#ff4596");
+	folder_colors["gray"] = Color("#9d9d9d");
```
Comment (Calinou) [inline editor/filesystem_dock.cpp:3436]: I suggest converting these `Color()` to their decimal form, as this will reduce binary size slightly (no more string to compile in the binary):

 ```suggestion
	folder_colors["red"] = Color(1.0, 0.271, 0.271);
	folder_colors["orange"] = Color(1.0, 0.561, 0.271);
	folder_colors["yellow"] = Color(1.0, 0.890, 0.271);
	folder_colors["green"] = Color(0.502, 1.0, 0.271);
	folder_colors["teal"] = Color(0.271, 1.0, 0.635);
	folder_colors["blue"] = Color(0.271, 0.843, 1.0);
	folder_colors["purple"] = Color(0.502, 0.271, 1.0);
	folder_colors["pink"] = Color(1.0, 0.271, 0.588);
	folder_colors["gray"] = Color(0.616, 0.616, 0.616);
```

Comment (the-sink) [reply editor/filesystem_dock.cpp]: Gotcha, updated in 5347641!

**Comment by the-sink** [comment]
Fixed the light theme folder icons:
image


You can see in the screenshot I also used the CanvasItem icon (felt it was the most suitable) for the option in the context menu. We could also leave it blank/without an icon, although I feel that'd be a little out of place considering all of the other items do have icons.

**Comment by DennisGHUA** [comment]
What's the reason behind certain .json files inheriting the color of their parent folder, while others do not? Isn't the expected behavior for all child folders and files to use the parent's color? Or did you set the color of those files to "default"?

**Comment by the-sink** [comment]
@DennisGHUA Oops, didn't notice that! Files will only inherit their parent folder color if the parent itself has a custom color set (not if it was inherited). I'll need to fix that.

**Comment by the-sink** [comment]
image

Seems to inherit properly now. It's worth mentioning that only a folder with a custom color set will have a lighter background color to distinguish it from child items; not every expanded folder with children will have a lighter hue (for instance, look at `some_folder67`). Personally I think this is fine, as the only alternative I could think of is progressively darkening the color of child folders as you descend in folder depth, which sounds like a bit of a clunky solution.

**Comment by Norrox** [comment]
I still think the more saturated background looks the best, if we are just going to get something in and tweak the colors later then I vote for the more colorful version to be merged, then we really see if people have problems with it 🙂
image

**Comment by TisFoolish** [comment]
What about a slider for saturation/opacity?

**Comment by DennisGHUA** [comment]
The PR looks good to me at its current state. Perhaps in the future, someone might suggest adding a saturation/opacity slider. If this PR gets merged and gains traction among users, it could spark fresh ideas.

However, it's important to note that if the memory issue isn't resolved, this PR might not be merged. I'm not well-versed in the godot source code, so my ability to assist with fixing it is limited. Interestingly, a previous commit did manage to pass the automated tests. It could be beneficial to compare the changes between then and now to pinpoint the source of the memory leak.

**Comment by the-sink** [comment]
I agree, it's probably worth adding saturation/opacity adjustments if there is a split between those who prefer the current vs the old, more visible colors, however I feel like we should try to get this PR out first so the feature is available. A separate PR can add further customizations/adjustments if needed.

I'm still unsure of what's causing the failed CI and it seems to be unrelated to the changes made here? Could be wrong though.

**Comment by Calinou** [comment]
> I'm still unsure of what's causing the failed CI and it seems to be unrelated to the changes made here? Could be wrong though.

You can rebase your fork against the latest upstream `master` branch and force-push so that CI is up-to-date.

**Comment by KoBeWi** [comment]
There are errors:

https://github.com/godotengine/godot/assets/2223172/1f940521-f273-47d3-b1f5-054e1e4f0e12

`clear()` does not remove submenu. You also enabled `set_unparent_when_invisible()`. The menu may be added multiple times or vanish too early.

File: editor/filesystem_dock.cpp [hunk 7]
```diff
@@ -182,16 +182,40 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
...
 	TreeItem *subdirectory_item = tree->create_item(p_parent);
 	String dname = p_dir->get_name();
+	String lpath = p_dir->get_path();
+
 	if (dname.is_empty()) {
 		dname = "res://";
 	}
 
+	// Set custom folder color (if applicable).
+	Dictionary custom_folder_colors = ProjectSettings::get_singleton()->get_setting("file_customization/folder_colors");
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:192]: `custom_folder_colors` can be a member of FileSystemDock instead. It would allow to use `get_setting()` once, in the constructor, instead of for every folder.

File: editor/filesystem_dock.cpp [hunk 8]
```diff
@@ -182,16 +182,40 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
...
+	bool has_custom_color = custom_folder_colors.has(lpath);
+	Color custom_color = has_custom_color ? folder_colors[custom_folder_colors[lpath]] : Color();
+
+	if (has_custom_color) {
+		bool is_dark_theme = EditorSettings::get_singleton()->is_dark_theme();
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:197]: This value can also (and should) be cached.

**Comment by Mickeon** [comment]
> `clear()` does not remove submenu. You also enabled `set_unparent_when_invisible()`. The menu may be added multiple times or vanish too early.

Could this be the cause of the memory leak?

**Comment by KoBeWi** [comment]
Subfolders in split view don't inherit color. Is this intended?

image

> Could this be the cause of the memory leak?

Yes.

File: editor/filesystem_dock.cpp [hunk 9]
```diff
@@ -2705,6 +2741,53 @@ void FileSystemDock::_get_drag_target_folder(String &target, bool &target_favori
...
+		String fpath = selected[i];
+
+		if (chosen_color_name) {
+			settings[fpath] = chosen_color_name;
+		} else {
+			settings.erase(fpath);
+		}
+	}
+
+	ProjectSettings::get_singleton()->set_setting("file_customization/folder_colors", settings);
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:2775]: Dictionaries are passed by reference, so this shouldn't be necessary.

File: editor/filesystem_dock.cpp [hunk 10]
```diff
@@ -2928,6 +3033,7 @@ void FileSystemDock::_file_list_item_clicked(int p_item, const Vector2 &p_pos, M
 	if (p_mouse_button_index != MouseButton::RIGHT) {
 		return;
 	}
+	files->grab_focus();
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:3036]: Same.

File: editor/filesystem_dock.cpp [hunk 11]
```diff
@@ -2881,6 +2984,8 @@ void FileSystemDock::_tree_rmb_select(const Vector2 &p_pos, MouseButton p_button
 	if (p_button != MouseButton::RIGHT) {
 		return;
 	}
+	tree->grab_focus();
```
[the remark this answers is not in the document]
Comment (the-sink) [reply editor/filesystem_dock.cpp:2987]: In my testing, right clicking tree items or file list items won't grab focus (left clicks do; right clicks don't). This feels unintentional, though, so it might be worth a separate PR to fix.

Behavior on the master branch:

https://github.com/godotengine/godot/assets/18225391/7314983a-cf91-44ec-9e69-18764ef3aeab

[the remark this answers is not in the document]
Comment (KoBeWi) [reply editor/filesystem_dock.cpp:2987]: Indeed, right clicks don't grab focus, see #75062
Though why is the focus needed here? Colors are changed for selected items, focus is irrelevant (?).

[the remark this answers is not in the document]
Comment (the-sink) [reply editor/filesystem_dock.cpp:2987]: I connected `id_pressed` for the folder colors submenu to its own method (`_folder_color_index_pressed`) instead of having `_file_option` handle it because I didn't want to fill up the FileMenu enum with options for all of the folder color items & end up having to hardcode the folder color submenu items (as opposed to letting it create them from the HashMap items, as it currently does). Although, that means it never receives the list of selected items like `_file_option` does, so I made it grab the list of selected items itself, and it needs which of the two menus are currently in focus for that.

Maybe I could write it so that it works similar to how the "New" submenu already works? I thought of doing that at first, but was hesitant to manually create all of the submenu options when it seemed more straightforward to do so by iterating the HashMap and checking the submenu item metadata in `_folder_color_index_pressed` instead.

[the remark this answers is not in the document]
Comment (KoBeWi) [reply editor/filesystem_dock.cpp:2987]: You could add only 2 enum values: `FOLDER_COLOR_DEFAULT` and `FOLDER_COLOR_MAX`, then instead of checking the `p_option` inside `switch` do `if (p_option >= FOLDER_COLOR_DEFAULT && p_option < FOLDER_COLOR_MAX)`. It's a bit hacky, but IMO acceptable. You could also add a comment that this is to avoid adding enum values for each color.

[the remark this answers is not in the document]
Comment (the-sink) [reply editor/filesystem_dock.cpp:2987]: Trying this out now - for the folder colors submenu to target the correct selections, I would need to connect `id_pressed` to either `_tree_rmb_option` or `_file_list_rmb_option` depending on which menu is in focus/was right-clicked anyways, wouldn't I? Because of that (as far as I can tell), the Create New submenu item doesn't even utilize the `p_selection` passed to `_file_option`, it uses the `current_path` variable of `FileSystemDock`, because it's always bound to `_tree_rmb_option` and wouldn't recieve the correct list of selections if this action was done on an item in the file list instead of the tree.

Perhaps I could use the existing system and look at `current_path` as the Create New submenu does, instead of checking focus and collecting selections as I'm doing now? It would mean only being able to set a folder color for one folder at a time, but I'm not sure if there are many scenarios where assigning many folders to the same color at a time would even be applicable, especially considering colors are inherited. Wouldn't have to mess with the enum, either.

[the remark this answers is not in the document]
Comment (KoBeWi) [reply editor/filesystem_dock.cpp:2987]: I can't believe a context menu is causing so many problems xd
I think restricting the option to single folder is fine. As you mentioned, folders inheriting colors make it a non-problem.

I think your `grab_focus()` solution could be fine too, but it changes behavior, so it needs opinion from more people (I don't think we grab focus on menus anywhere else in the editor). You could add a member `Control *context_menu_owner` set when calling `_file_and_folders_fill_popup()` and use it instead of checking focus.

**Comment by the-sink** [comment]
> Subfolders in split view don't inherit color. Is this intended?

Definitely not - seems that anything more than one folder depth into a folder with a custom color assigned will not inherit the parent color, will fix that.

> Also setting a color in split view will navigate into the recolored directory.

This is definitely weird behavior, but seems to be a larger quirk with the filesystem dock. Doing any action in the right-click context menu will assign `current_path` to the item that you've right-clicked, and as such any time the file list is redrawn it will render the contents of that folder. A good example of this behavior would be right clicking a folder in the file list, choosing "Copy Path", and then forcing a redraw by toggling the file list's grid view in the top right of the menu.


https://github.com/godotengine/godot/assets/18225391/ae236797-198b-41c8-9891-b2929841117b

This also causes the file list view to navigate into a folder after it has been renamed, which feels unintuitive (to me, at least). Worth a separate github issue maybe?

File: editor/filesystem_dock.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (the-sink) [reply editor/filesystem_dock.cpp:2999]: It might be beneficial to leave the `grab_focus()` solution in for now until #75062 is addressed, as it also fixes the rename issue (#78789). If it should be removed, I'll probably go ahead and use `current_path` instead to make it a little bit more straightforward.

**Comment by the-sink** [comment]
[451c7d3](https://github.com/godotengine/godot/commit/451c7d3358d0aee0ab1aec3ffac49b18290fc4c7) passed CI but [c682745](https://github.com/godotengine/godot/commit/c682745614f4d0154883aa9d96a3b3457d351e32) didn't, despite the only changes being related to comments...?

Otherwise, I *believe* pretty all of the issues that were brought up have been addressed.

File: editor/filesystem_dock.cpp [hunk 12]
```diff
@@ -2797,6 +2891,27 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, Vector<Str
 		}
 
 		p_popup->add_separator();
+
+		if (p_paths[0] != "res://") {
+			PopupMenu *folder_colors_menu = memnew(PopupMenu);
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:2896]: This isn't correct either, the menu is not freed automatically, so you are just making infinite menus (and only the first one is used), see #79965

Comment (the-sink) [reply editor/filesystem_dock.cpp:2896]: Would it be sufficient for now to create the folder colors submenu in the constructor & parent it to the tree PopupMenu initially, but `reparent()` it when it needs to be moved?

Comment (KoBeWi) [reply editor/filesystem_dock.cpp:2896]: The PR I linked got approved and might be merged soon, so it will resolve this issue. It's probably fine to leave it as is right now, sorry for bother >_>

Comment (the-sink) [reply editor/filesystem_dock.cpp:2896]: No worries haha, that works.

File: editor/filesystem_dock.cpp [hunk 13]
```diff
@@ -899,7 +939,7 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
 	}
 
 	Ref<Texture2D> folder_icon = (use_thumbnails) ? folder_thumbnail : get_theme_icon(SNAME("folder"), SNAME("FileDialog"));
-	const Color folder_color = get_theme_color(SNAME("folder_icon_color"), SNAME("FileDialog"));
+	Color default_folder_color = get_theme_color(SNAME("folder_icon_color"), SNAME("FileDialog"));
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:942]: Why did you remove `const`?

**Review by KoBeWi — APPROVED** [review]

File: editor/filesystem_dock.cpp (hunk 13 above)
Comment (the-sink) [reply editor/filesystem_dock.cpp:942]: Oops, at one point I was changing that variable further down in order to test something afaik but must've forgotten to make it `const` again.

**Comment by akien-mga** [comment]
Needs rebase after #61818, you should now use the new `ProjectSettings::add_hidden_prefix` to hide the prefixed properties.

**Comment by akien-mga** [comment]
Thanks!

File: editor/filesystem_dock.cpp (hunk 9 above)
Comment (AThousandShips) [reply editor/filesystem_dock.cpp]: Not necessarily true here, as if the dictionary is *empty* which it is when starting the settings will be returned as a default value when fetched and not properly tracked

Comment (KoBeWi) [reply editor/filesystem_dock.cpp]: Right, but if the setting existed, this would do a redundant set.

Comment (AThousandShips) [reply editor/filesystem_dock.cpp]: Sure, but the setting doesn't exist for any new user, and won't ever unless they already had it, which they won't

**Comment by JekSun97** [comment]
This is needed in 3.6+

**Comment by Calinou** [comment]
> This is needed in 3.6+

To my knowledge, this PR relies on Tree changes only found in 4.0 (double-check this though). As a result, if this turns out to be the case, it can't be backported easily for 3.6.

File: editor/filesystem_dock.cpp [hunk 14]
```diff
@@ -2798,6 +2892,27 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, Vector<Str
...
+			p_popup->add_child(folder_colors_menu);
+			p_popup->add_submenu_item(TTR("Set Folder Color..."), "FolderColor");
+			p_popup->set_item_icon(-1, get_theme_icon(SNAME("CanvasItem"), SNAME("EditorIcons")));
+
+			folder_colors_menu->add_icon_item(get_theme_icon(SNAME("Folder"), SNAME("EditorIcons")), TTR("Default (Reset)"));
+			folder_colors_menu->set_item_icon_modulate(0, get_theme_color(SNAME("folder_icon_color"), SNAME("FileDialog")));
+			folder_colors_menu->add_separator();
+
+			for (const KeyValue<String, Color> &E : folder_colors) {
+				folder_colors_menu->add_icon_item(get_theme_icon(SNAME("Folder"), SNAME("EditorIcons")), TTR(E.key.capitalize()));
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:2910]: `TTR(E.key.capitalize()))`
This is wrong. The color names need to be provided as explicit strings.

Comment (Mickeon) [reply editor/filesystem_dock.cpp:2910]: Oh yeah, you absolutely cannot translate that. It needs to be addressed ASAP.

**Comment by GuyUnger** [comment]
It would be very useful if the colors would show up in more places, made a PR for it https://github.com/godotengine/godot-proposals/issues/8581

**Comment by Splendid-Failuers** [comment]
Would be really nice to have this extend to all icons, and not just folders. I've used colours to track state of files across a project before in other engines, and really miss that feature in Godot. Setting a red icon for a file if it's a block-up asset, yellow if it's in review, and green if it's a final asset proposal. Really nice to not have to track progress on all files in an external program.
